src/roles/actors.rs

```diff
@@ -0,0 +1,356 @@
+use srml_support::traits::{Currency, EnsureAccountLiquid};
+use runtime_primitives::traits::{Zero, Bounded, SimpleArithmetic, As, Member, MaybeSerializeDebug, MaybeDebug};
+use system::{self, ensure_signed};
+use crate::governance::{GovernanceCurrency, BalanceOf };
+
+use crate::traits::{Members, Roles};
+
+#[derive(Encode, Decode, Copy, Clone, Eq, PartialEq, Debug)]
+pub enum Role {
+    Storage,
+}
+
+
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Copy, Clone, Eq, PartialEq)]
+pub struct RoleParameters<T: Trait> {
+    // minium balance required to stake to enter a role
+    pub min_stake: BalanceOf<T>,
+
+    // the maximum number of spots available to fill for a role
+    pub max_actors: u32,
+
+    // minimum actors to maintain - if role is unstaking
+    // and remaining actors would be less that this value - prevent or punish for unstaking
+    pub min_actors: u32,
+
+    // fixed amount of tokens paid to actors' primary account
+    pub reward: BalanceOf<T>,
+
+    // payouts are made at this block interval
+    pub reward_period: T::BlockNumber,
+
+    // how long tokens remain locked for after unstaking
+    pub unbonding_period: T::BlockNumber,
+
+    // minimum amount of time before being able to unstake
+    pub bonding_time: T::BlockNumber,
+
+    // minimum period required to be in service. unbonding before this time is highly penalized
+    pub min_service_period: T::BlockNumber,
+
+    // "startup" time allowed for roles that need to sync their infrastructure
+    // with other providers before they are considered in service and punishable for
+    // not delivering required level of service.
+    pub startup_grace_period: T::BlockNumber,
+
+    // small fee burned to make a request to enter role
+    pub entry_request_fee: BalanceOf<T>,
+}
+
+#[derive(Encode, Decode, Clone)]
+pub struct Actor<T: Trait> {
+    member_id: MemberId<T>,
+    role: Role,
+    account: T::AccountId,
+    joined_at: T::BlockNumber,
+    // startup_grace_period_ends_at: T::BlockNumber,
+}
+
+pub trait Trait: system::Trait + GovernanceCurrency + MaybeDebug {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    type Members: Members<Self>;
+}
+
+pub type MemberId<T: Trait> = <T::Members as Members<T>>::Id;
+pub type Request<T: Trait> = (T::AccountId, MemberId<T>, Role, T::BlockNumber); // actor account, memberid, role, expires
+pub type Requests<T: Trait> = Vec<Request<T>>;
+
+const REQUEST_LIFETIME: u64 = 300;
+const DEFAULT_REQUEST_CLEARING_INTERVAL: u64 = 100;
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Actors {
+        /// requirements to enter and maintain status in roles
+        Parameters get(parameters) : map Role => Option<RoleParameters<T>>;
+
+        /// the roles members can enter into
+        AvailableRoles get(available_roles) : Vec<Role>; // = vec![Role::Storage];
+
+        /// Actors list
+        Actors get(actors) : Vec<T::AccountId>;
```

Comment:
Actually, this is a list of actors' account ids, not Actors
```suggestion
        ActorIds get(actor_ids) : Vec<T::AccountId>;
```